src/libpmemobj/redo.c

```diff
@@ -322,6 +347,75 @@ redo_log_entry_val_create(struct redo_log *redo, size_t offset, uint64_t *dest,
 	return e;
 }
 
+/*
+ * redo_log_entry_buf_create -- atomically creates a buffer entry in the log
+ */
+struct redo_log_entry_buf *
+redo_log_entry_buf_create(struct redo_log *redo, size_t offset, uint64_t *dest,
+	const void *src, uint64_t size,
+	enum redo_operation_type type, const struct pmem_ops *p_ops)
+{
+	struct redo_log_entry_buf *e =
+		(struct redo_log_entry_buf *)(redo->data + offset);
+
+	/*
+	 * Depending on the size of the source buffer, we might need to perform
+	 * up to three separate copies:
+	 *	1. The first cacheline, 24b of metadata and 40b of data
+	 * If there's still data to be logged:
+	 *	2. The entire remainder of data data aligned down to cacheline,
+	 *	for example, if there's 150b left, this step will copy only
+	 *	128b.
+	 * Now, we are left with between 0 to 63 bytes. If nonzero:
+	 *	3. Create a stack allocated cacheline-sized buffer, fill in the
+	 *	remainder of the data, and copy the entire cacheline.
+	 *
+	 * This is done so that we avoid a cache-miss on misaligned writes.
+	 */
+
+	struct redo_log_entry_buf *b = alloca(CACHELINE_SIZE);
+	b->base.offset = (uint64_t)(dest) - (uint64_t)p_ops->base;
+	b->base.offset |= REDO_OPERATION(type);
+	b->size = size;
+
+	size_t ncopy = MIN(size,
+		CACHELINE_SIZE - sizeof(struct redo_log_entry_buf));
+	memcpy(b->data, src, ncopy);
+
+	size_t remaining_size = ncopy > size ? 0 : size - ncopy;
+
+	char *srcof = (char *)src + ncopy;
+	size_t rcopy = ALIGN_DOWN(remaining_size, CACHELINE_SIZE);
+	size_t lcopy = remaining_size - rcopy;
+
+	uint8_t last_cacheline[CACHELINE_SIZE];
+	if (lcopy != 0)
+		memcpy(last_cacheline, srcof + rcopy, lcopy);
+
+	b->checksum = util_checksum_seq(b, CACHELINE_SIZE, 0);
+	if (rcopy != 0)
+		b->checksum = util_checksum_seq(srcof, rcopy, b->checksum);
+	if (lcopy != 0)
+		b->checksum = util_checksum_seq(last_cacheline,
+			CACHELINE_SIZE, b->checksum);
+
+	pmemops_memcpy(p_ops, e, b, CACHELINE_SIZE,
+		PMEMOBJ_F_MEM_NODRAIN | PMEMOBJ_F_MEM_NONTEMPORAL);
+
+	if (rcopy != 0)
+		pmemops_memcpy(p_ops, e->data + ncopy, srcof, rcopy,
+			PMEMOBJ_F_MEM_NODRAIN | PMEMOBJ_F_MEM_NONTEMPORAL);
+
+	if (lcopy != 0)
+		pmemops_memcpy(p_ops, e->data + ncopy + rcopy, last_cacheline,
+			CACHELINE_SIZE,
+			PMEMOBJ_F_MEM_NODRAIN | PMEMOBJ_F_MEM_NONTEMPORAL);
+
+	pmemops_drain(p_ops);
```

Comment:
One drain per one redo log entry is going to hurt.

Reply:
Ok, it seems it's used only for CPY and SET operations (for now undo logs only), so it doesn't matter that much. But should we have this drain here or on the caller's side?

Reply:
I don't think so, but let's do this separately to this patch. Redo logs will probably use a simplified form of this function, without the drain.